app/uk/gov/hmrc/selfservicetimetopay/controllers/CalculatorController.scala

```diff
@@ -18,94 +18,302 @@ package uk.gov.hmrc.selfservicetimetopay.controllers
-      "dueByYear" -> number(min = 2000, max = 2100)
+      "dueByDay" -> number(min = 1, max = 31)
     )(CalculatorAmountDue.apply)(CalculatorAmountDue.unapply))
   }
 
-  private def createPaymentTodayForm: Form[CalculatorPaymentToday] = {
+
+  private def createPaymentTodayForm(totalDue:BigDecimal): Form[CalculatorPaymentToday] = {
     Form(mapping(
-      "amount" -> optional(bigDecimal)
+      "amount" -> optional(bigDecimal
+        .verifying("ssttp.calculator.form.payment_today.amount.less-than-owed", _ < totalDue)
+        .verifying("ssttp.calculator.form.payment_today.amount.nonnegitive", _.compare(BigDecimal("0")) >= 0)
+    )
     )(CalculatorPaymentToday.apply)(CalculatorPaymentToday.unapply))
   }
 
-  private def createDurationForm(min: Integer, max: Integer): Form[CalculatorDuration] = {
+  private def createDurationForm(min: Int, max: Int): Form[CalculatorDuration] = {
+    def required: Constraint[Int] = Constraint[Int]("constraint.required") { o =>
+      if (o == null) Invalid(ValidationError("ssttp.calculator.form.duration.months.required")) else Valid
+    }
+    def greaterThan: Constraint[Int] = Constraint[Int]("constraint.duration-more-than") { o =>
+      if (o != null && o < min) Invalid(ValidationError("ssttp.calculator.form.duration.months.greater-than", min)) else Valid
+    }
+    def lessThan: Constraint[Int] = Constraint[Int]("constraint.duration-more-than") { o =>
+      if (o != null && o > max) Invalid(ValidationError("ssttp.calculator.form.duration.months.less-than", max)) else Valid
+    }
     Form(mapping(
-      "months" -> number(min = min, max = max)
+      "months" -> number.verifying(required, greaterThan, lessThan)
     )(CalculatorDuration.apply)(CalculatorDuration.unapply))
   }
 
+  private def getKeystoreData(implicit request:Request[AnyContent]) : (Boolean, Option[CalculatorAmountsDue], Option[CalculatorPaymentToday], Option[CalculatorDuration], Option[Seq[CalculatorPaymentSchedule]]) = {
+    val ssttpStart = request.session.get("SelfServiceTimeToPayStart").isDefined
+    val amountsDue:Option[CalculatorAmountsDue] = request.session.get("CalculatorAmountsDue") match {
+      case Some(json) => Option(JacksonMapper.readValue(json, classOf[CalculatorAmountsDue]))
+      case None => None
+    }
+    val paymentToday:Option[CalculatorPaymentToday] = request.session.get("CalculatorPaymentToday") match {
+      case Some(json) => Option(JacksonMapper.readValue(json, classOf[CalculatorPaymentToday]))
+      case None => None
+    }
+    val duration:Option[CalculatorDuration] = request.session.get("CalculatorDuration") match {
+      case Some(json) => Option(JacksonMapper.readValue(json, classOf[CalculatorDuration]))
+      case None => None
+    }
+    val schedules:Option[Seq[CalculatorPaymentSchedule]] = request.session.get("CalculatorPaymentSchedules") match {
+      case Some(json) => Some(generatePaymentSchedules(amountsDue.get.total, paymentToday.get.amount))
```

Comment:
Calls to get ( amountsDue.get.total AND paymentToday.get.amount ) are not safe, if you know that options will always have a value then don't model the return type with option. Besides this the entire function is more verbose than necessary, there is no need to match on option especially if you're returning None for a None it would be better to use map.